Switches from Webclient to HttpClient

LukeMarlin opened the pull request:
- Features:
    - HttpClient can be passed as an argument to the constructor, making
    it possible to reuse an existing one
    - HttpClient instance is static, so all Ovh.Client share the same
    - Obsoletes all sync method, as they do not exist in HttpClient
    - Switches README to [...]Async methods
  - Fixes:
    - Character separator for Batch requests is no longer ignored
    - Timeout is no longer ignored (note: for now, and because
    the HttpClient instance is static, the value given to the first
    Ovh.Client instantiation will be used for all other instances)
  - Testing:
    - Exposes GenerateSignature
    - Exposes a way to mock DateTimeOffset
    - Adds a first round of tests for signature and calls

LukeMarlin committed:
Switches from Webclient to HttpClient

  - Features:
    - HttpClient can be passed as an argument to the constructor, making
    it possible to reuse an existing one
    - HttpClient instance is static, so all Ovh.Client share the same
    - Obsoletes all sync method, as they do not exist in HttpClient
    - Switches README to [...]Async methods
  - Fixes:
    - Character separator for Batch requests is no longer ignored
    - Timeout is no longer ignored (note: for now, and because
    the HttpClient instance is static, the value given to the first
    Ovh.Client instantiation will be used for all other instances)
  - Testing:
    - Exposes GenerateSignature
    - Exposes a way to mock DateTimeOffset
    - Adds a first round of tests for signature and calls

LukeMarlin commented:
Fixes #28

Namaneo commented on csharp-ovh/Client/Client.cs:
```diff
@@ -229,30 +195,113 @@ private Client()
+            else
+            {
+                _httpClient = httpClient;
+            }
```
Would it be a good idea to have two constructors instead?
* One that optionally takes a timeout as parameter and initialize an `HttpClient` (BTW the `Timeout` property assignation seems useless now?)
* One that directly takes an `HttpClient`

I don't know if it would be overkill, but at least we know for sure the purpose of the different parameters? Currently you can pass both a timeout and an client without suspecting that the timeout will be useless.
(As a side note, `LoadConfiguration` could even be a private constructor)

LukeMarlin replied:
Makes sense to have two of them, yes.
Regarding the timeout, it will be changed in with #32 which will rework all mechanisms linked to it. You can see it's been obsoleted already.

Namaneo commented on csharp-ovh/Client/Client.cs:
```diff
@@ -229,30 +195,113 @@ private Client()
+        {
+            _timeProvider = customTimeProvider;
+            return this;
+        }
```
If this method should not be used for production code, why don't you make it internal?
If you do, then you can tell in your csproj that you test project is a friend assembly :
```xml
<ItemGroup>
  <AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
    <_Parameter1>Your.Test.Assembly</_Parameter1>
  </AssemblyAttribute>
</ItemGroup>
```

LukeMarlin replied:
I was hesitating to leave it as `public` to let user find usages for it. Might not be a good idea though, so yeah I moved it to `internal`

Namaneo commented on csharp-ovh/Client/Client.cs:
```diff
@@ -310,23 +356,7 @@ private WebHeaderCollection GetHeaders(string method, string data, bool needAuth
-            {
-                throw HandleWebException(ex);
-            }
+            return CallAsync(method, path, data, needAuth, isBatch).Result;
```
If `Call` is only intended to be used synchronously, you may use `.GetAwaiter().GetResult()` instead, as this does not wrap exceptions into an `AggregateException`.
(I know I know, this is deprecated, but if you accidentally go through this method while fixing stuff, you could change it as well?)

EDIT: Also applies (without deprecation this time) to `TimeDelta`

LukeMarlin replied:
As per [the doc](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.taskawaiter?view=netframework-4.8), these methods are intended to be used by compiler only, so I didn't use them at first. However, I noticed that the move to `.Result` is a breaking change because of the `AggregateException`... So I changed it to `GetResult` anyway to keep the compatibility.

Namaneo commented on README.rst:
```diff
@@ -18,7 +18,7 @@ credential creation and requests signing.
                 // to get your credentials
                 Client client = new Client("ovh-eu", "<application_key>", "<application_secret>", "<consumer_key>");
-                PartialMe me = client.Get<PartialMe>("/me");
+                PartialMe me = client.GetAsync<PartialMe>("/me").Result;
```
As of [C# 7.1](https://docs.microsoft.com/fr-fr/dotnet/csharp/whats-new/csharp-7-1), you can have async `Main`.
Maybe this is a good opportunity to update the doc with the `async/await` mechanism?

LukeMarlin replied:
Yes, good point

Namaneo commented on csharp-ovh/Client/Client.cs:
```diff
@@ -265,27 +314,24 @@ private WebHeaderCollection GetHeaders(string method, string data, bool needAuth
-                headers.Add("X-Ovh-Signature", "$1$" + signature);
+                headers.Add("X-Ovh-Timestamp", currentTimestamp.ToString());
+                headers.Add("X-Ovh-Signature", signature);
             }
```
Pretty long `if` statement : maybe put `isBatch` check before, and have the opposite condition right after?
```csharp
if (!needAuth)
{
  return;
}

// ... The auth stuff ...
```

Namaneo commented on csharp-ovh/Client/Client.cs:
```diff
@@ -51,31 +50,15 @@ namespace Ovh.Api
     public partial class Client
     {
         private readonly Dictionary<string, string> _endpoints =
             new Dictionary<string, string>();
```
Maybe move to `Dictionary<string, Uri>`?

LukeMarlin replied:
I could, but what would be the benefit?

Namaneo commented on csharp-ovh/Client/Client.cs:
```diff
@@ -347,60 +377,40 @@ private string Call(string method, string path, string data = null, bool needAut
-                    return new HttpException("Low HTTP request failed error", ex);
-                }
-                return ExtractExceptionFromHttpError(httpResponse, ex);
+                throw new HttpException("An exception occured while trying to issue the HTTP call", e);
```
Why not directly throwing the `HttpRequestException`?

LukeMarlin replied:
Previously, to stick with the exception names and behaviour of python-ovh.
Now, because it would be a breaking change. I might do it for 4.0 though...

Namaneo commented on test/Constants.cs:
```diff
@@ -0,0 +1,16 @@
+        public const string OVH_TIME_HEADER = "X-Ovh-Timestamp";
+        public const string OVH_SIGNATURE_HEADER = "X-Ovh-Signature";
+        public const string OVH_BATCH_HEADER = "X-Ovh-Batch";
+    }
```
You can move this constants class directly on the `Client` part of the code : this may improve consistency between production code and tests?

LukeMarlin replied:
Yeah, I probably started by the testing ones, then added these without thinking. Changed

Namaneo commented on test/test.csproj:
```diff
@@ -1,17 +1,15 @@
+    <PackageReference Include="FakeItEasy" Version="5.1.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.3.0" />
+    <PackageReference Include="NUnit" Version="3.8.1" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.8.0" />
```
Just wondering why NUnit over Xunit?

LukeMarlin replied:
At the time I wrote the lib, it seemed that xUnit was a bit young and underdocumented. Might be wrong though. If I were to restart I'd use xUnit, I might even do it later on... So no particular reason ;)

Namaneo commented on test/DeleteRequests.cs:
```diff
@@ -0,0 +1,82 @@
+            queryParams.Add("filter", "value:&é'-");
+            queryParams.Add("anotherfilter", "=test");
+            var result = c.DeleteAsync<object>("/ip/127.0.0.1").Result;
+            Assert.IsNull(result);
```
If result of `DeleteAsync` should be null, what's the purpose of providing the generic version of this method?

LukeMarlin replied:
Might not be `null`. Ovh's API does not enforce to return `null` on DELETE calls.

Namaneo commented on test/GetRequests.cs:
```diff
@@ -0,0 +1,125 @@
+
+
+            var c = ClientFactory.GetClient(fake);
+            var result = c.GetAsync("/me").Result;
```
Global thought about test methods : it seems that NUnit [supports async methods](https://stackoverflow.com/questions/12191831/how-do-i-test-an-async-method-with-nunit-or-possibly-with-another-framework), would you give it a try?

LukeMarlin replied:
Yup, done. Not sure why I didn't before :astonished:

Namaneo commented on test/PutRequests.cs:
```diff
@@ -0,0 +1,101 @@
+                Assert.AreEqual(Constants.CONSUMER_KEY, headers.GetValues(Constants.OVH_CONSUMER_HEADER).First());
+                Assert.AreEqual(currentServerTimestamp.ToString(), headers.GetValues(Constants.OVH_TIME_HEADER).First());
+                Assert.AreEqual("$1$5e81842c0f0c806fd703de084d80192a59bc0f8a", headers.GetValues(Constants.OVH_SIGNATURE_HEADER).First());
+            });
```
Would an extension method on `headers` (or even a simple static one) be a good idea? Even something dirty where you pass headers, HTTP method, signature ans timestamp : this way would prevent code duplication, and you will be consistent on those checks (as they seem to be the same for each route)

LukeMarlin replied:
I did it before, but that was hiding the tests in another method and I wasn't sure I liked it so I removed it.
Maybe unitary methods like `AssertConsumerKeyHeaderPresent`, but that's for another time.

Namaneo commented on test/Responses/DeleteResponses.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        public static string nullAsJsonString = "null";
+        public static HttpResponseMessage nullAsHttpResponseMessage = new HttpResponseMessage{
+            Content = new StringContent(nullAsJsonString, Encoding.UTF8, "application/json"),
```
Could `"null"` be inlined here? Or at least private?
Could both be readonly?
Could `new HttpResponseMessage { ... }` be in a static method to avoid code duplication?

(Note: may apply to other response classes)

LukeMarlin replied:
- Might be, but I might refer to it from future tests and I'm not sure that it would be a big perf improvement anyway :)
- Yes, done
- Done

Namaneo commented on test/Responses/Models/Address.cs:
```diff
@@ -0,0 +1,11 @@
+        public string country { get; set; }
+        public string line1 { get; set; }
+        public string zip { get; set; }
+    }
```
You're defining real (?) return values here : is there any reason not to provide those routes directly inside the client?

LukeMarlin replied:
- While they work, they are only the required properties for requests, and some selected properties for responses, so they wouldn't fit
- The lib is not intended to be a full fledged SDK. It will never embark Models as this requires careful monitoring of API changes

LukeMarlin committed:
Changes according to Julien Loir's review:

  - Splits constructor in two so that arguments are clearer
  - Makes .AsTestable internal and declares test assembly as friend
  - Switches from .Result to .GetAwaiter().GetResult() to avoid breaking
  change on the exceptions that the lib can send
  - Updates README to use async main in examples
  - Reverses condition in SetHeaders for readability
  - Makes OVH header names available as constants in Client
  - Makes tests async
  - Adds Ovh.Test.HttpResponseMessageFactory to avoid code duplication
  in test cases

Additional stuff:
  - Makes the _httpClient instance static for real
  - Correct obsolete statements
  - Auto styling

Namaneo commented:
All good for me since recent updates.
LGTM
